dali/kernels/imgproc/jpeg/jpeg_distortion_gpu.cuh

```diff
@@ -174,34 +239,52 @@ __global__ void ChromaSubsampleDistortion(const SampleDesc *samples,
   }
 }
 
+__device__ __inline__ float quantize(float value, float Q_coeff) {
+  if (Q_coeff < 1)
+    Q_coeff = 1;
+  return Q_coeff * roundf(value * __frcp_rn(Q_coeff));
+}
 
-
-template <bool horz_subsample, bool vert_subsample>
+/**
+ * @brief Produces JPEG compression artifacts by running the lossy part of
+ * JPEG compression and decompression.
+ */
+template <bool horz_subsample, bool vert_subsample, bool quantization = true>
 __global__ void JpegCompressionDistortion(const SampleDesc *samples,
                                           const kernels::BlockDesc<2> *blocks) {
   using T = uint8_t;
   const auto &block = blocks[blockIdx.x];
   const auto &sample = samples[block.sample_idx];
+  const auto &luma_Q_table = sample.luma_Q_table;
+  const auto &chroma_Q_table = sample.chroma_Q_table;
 
-  int y_start = threadIdx.y + block.start.y;
-  int x_start = threadIdx.x + block.start.x;
-  if (y_start >= block.end.y || x_start >= block.end.x) {
-    return;
-  }
+  static constexpr int align_y = 8;
+  static constexpr int align_x = 8;
+  int aligned_start_x = block.start.x & -align_x;  // align down
+  int aligned_start_y = block.start.y & -align_y;
+  int aligned_end_y = align_up(block.end.y, align_y);
+  int aligned_end_x = align_up(block.end.x, align_x);
 
   // Assuming CUDA block is 32x16, leading to a 2x4 grid of chroma blocks
   // and up to 4x8 blocks of luma.
-  __shared__ T luma_blk[2 << vert_subsample][4 << horz_subsample][8][9];
-  __shared__ T cb_blk[2][4][8][9];  // 8+1 to reduce bank conflicts
-  __shared__ T cr_blk[2][4][8][9];  // 8+1 to reduce bank conflicts
+  const int vert_chroma_blocks = 2;
+  const int vert_luma_blocks = vert_chroma_blocks << vert_subsample;
+  const int horz_chroma_blocks = 4;
+  const int horz_luma_blocks = horz_chroma_blocks << horz_subsample;
```

Comment:
So what we can do is:
- 444
- 422
- 420
- 440

But not 411, right?

Reply:
We can do 2x2, 1x2, 2x1 and 1x1 subsampling.
I don't know how these translate it to these names (and if they even can describe the difference between 1x2 and 2x1).

Reply:
https://en.wikipedia.org/wiki/Chroma_subsampling#Types_of_sampling_and_subsampling
They cannot express 1xN.

Reply:
444 : horz=false, vert=false
422 : horz=true, vert=false
420 : horz=true, vert=true
440 : horz=false, vert=true

411 : not supported, that is subsampled with factor 4 horizontally